aptos-move/framework/aptos-framework/sources/account/rate_limiter.move

```diff
@@ -0,0 +1,166 @@
+module aptos_framework::rate_limiter {
+    use aptos_std::math64;
+    use aptos_framework::timestamp;
+
+    enum RateLimiter has key, store, copy, drop {
+        // Struct to represent a Token Bucket that refills every minute
+        TokenBucket {
+            // Maximum number of tokens allowed at any time.
+            capacity: u64,
+            // Current number of tokens remaining in this interval.
+            current_amount: u64,
+            // refill `capacity` number of tokens every `refill_interval` in seconds.
+            refill_interval: u64,
+            // Last time the bucket was refilled (in seconds)
+            last_refill_timestamp: u64,
+            // accumulated amount that hasn't yet added up to a full token
+            fractional_accumulated: u64,
+        }
+    }
+
+    // Public entry function to initialize a Token Bucket based rate limiter.
+    public fun initialize(capacity: u64, refill_interval: u64): RateLimiter {
+        RateLimiter::TokenBucket {
+            capacity,
+            current_amount: capacity, // Start with a full bucket (full capacity of transactions allowed)
+            refill_interval,
+            last_refill_timestamp: timestamp::now_seconds(),
+            fractional_accumulated: 0, // Start with no fractional accumulated
+        }
+    }
+
+    // Public function to request a transaction from the bucket
+    public fun request(limiter: &mut RateLimiter, num_token_requested: u64): bool {
+        refill(limiter);
+        if (limiter.current_amount >= num_token_requested) {
+            limiter.current_amount = limiter.current_amount - num_token_requested;
+            true
+        } else {
+            false
+        }
+    }
+
+    // Function to refill the transactions in the bucket based on time passed
+    fun refill(limiter: &mut RateLimiter) {
+        let current_time = timestamp::now_seconds();
+        let time_passed = current_time - limiter.last_refill_timestamp;
+        // Calculate the full tokens that can be added
+        let accumulated_amount = time_passed * limiter.capacity + limiter.fractional_accumulated;
+        let new_tokens = accumulated_amount / limiter.refill_interval;
+        // Calculate the remaining fractional amount.
+
+        if (new_tokens > 0) {
+            limiter.current_amount = math64::min(limiter.current_amount + new_tokens, limiter.capacity);
+            limiter.last_refill_timestamp = current_time;
+        };
+
+        // Update the fractional amount accumulated for the next refill cycle
+        limiter.fractional_accumulated = accumulated_amount % limiter.refill_interval;
```

Comment:
When rate limiter exceeds capacity, will the remainder added to this fractional_accumulated?

Reply:
good point.
if limiter.current_amount + new_tokens < limiter.capacity, we 
`limiter.fractional_accumulated = accumulated_amount % limiter.refill_interval;`
otherwise it should be 0
cc: @igor-aptos